modules/navier_stokes/src/physics/WCNSFVScalarAdvectionPhysics.C

```diff
@@ -0,0 +1,385 @@
+{
+  const std::string kernel_type = "FVCoupledForce";
+  InputParameters params = getFactory().getValidParams(kernel_type);
+  assignBlocks(params, _blocks);
+
+  for (const auto scalar_i : index_range(_passive_scalar_names))
+  {
+    params.set<NonlinearVariableName>("variable") = _passive_scalar_names[scalar_i];
+    if (_passive_scalar_sources.size())
+    {
+      // Added for backward compatibility with former Modules/NavierStokesFV syntax
+      params.set<MooseFunctorName>("v") = _passive_scalar_sources[scalar_i];
+      getProblem().addFVKernel(
+          kernel_type, prefix() + "ins_" + _passive_scalar_names[scalar_i] + "_source", params);
+    }
+
+    // Sufficient for all intents and purposes
+    if (_passive_scalar_coupled_sources.size())
+      for (const auto i : index_range(_passive_scalar_coupled_sources[scalar_i]))
+      {
+        params.set<MooseFunctorName>("v") = _passive_scalar_coupled_sources[scalar_i][i];
+        if (_passive_scalar_sources_coef.size())
+          params.set<Real>("coef") = _passive_scalar_sources_coef[scalar_i][i];
+
+        getProblem().addFVKernel(kernel_type,
+                                 prefix() + "ins_" + _passive_scalar_names[scalar_i] +
+                                     "_coupled_source_" + std::to_string(i),
+                                 params);
+      }
+  }
+}
+
+void
+WCNSFVScalarAdvectionPhysics::addFVBCs()
+{
+  // For compatibility with Modules/NavierStokesFV syntax
+  if (!_has_scalar_equation)
+    return;
+
+  addScalarInletBC();
+  // There is typically no wall flux of passive scalars, similarly we rarely know
+  // their concentrations at the outlet at the beginning of the simulation
+  // TODO: we will know the outlet values in case of flow reversal. Implement scalar outlet
+}
+
+void
+WCNSFVScalarAdvectionPhysics::addScalarInletBC()
+{
+  const auto & inlet_boundaries = _flow_equations_physics->getInletBoundaries();
+
+  // Boundary checks
+  // TODO: once we have vectors of MooseEnum, we could use the same templated check for types and
+  // functors
+  if (inlet_boundaries.size() * _passive_scalar_names.size() != _passive_scalar_inlet_types.size())
+    paramError(
+        "passive_scalar_inlet_types",
+        "The number of scalar inlet types (" + std::to_string(_passive_scalar_inlet_types.size()) +
+            ") is not equal to the number of inlet boundaries (" +
+            std::to_string(inlet_boundaries.size()) + ") times the number of passive scalars (" +
+            std::to_string(_passive_scalar_names.size()) + ")");
+  if (_passive_scalar_names.size() != _passive_scalar_inlet_functors.size())
+    paramError("passive_scalar_inlet_functors",
+               "The number of groups of inlet functors (" +
+                   std::to_string(_passive_scalar_inlet_functors.size()) +
+                   ") is not equal to the number of passive scalars (" +
+                   std::to_string(_passive_scalar_names.size()) + ")");
```

Comment:
Maybe you can replace this with you new parameter checks.

Reply:
components havent had the chance to modify the boundaries at this point
so we must wait till last minute, and cant rely on parameters only